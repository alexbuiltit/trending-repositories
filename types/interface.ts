export interface CustomError extends Error {
  displayMessage?: string;
}

export interface ErrorHandling {
  success?: boolean;
  error?: CustomError;
}
