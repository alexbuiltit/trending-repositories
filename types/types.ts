export type RepositoryType = {
  name: string;
  description?: string | null;
  starCount: number;
  id: number;
  language?: string | null;
  link: string;
  favourite?: boolean;
  toggleFavourite?: (id: number) => void;
};

export type RepositoryListType = {
  items: RepositoryType[] | undefined;
  filters?: string[] | undefined;
  toggleFavourite?: (id: number) => void;
  favourites?: number[] | undefined;
};
