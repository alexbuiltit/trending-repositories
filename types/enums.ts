export enum ListingCategory {
  All = 'all',
  Favourites = 'favourites',
}
