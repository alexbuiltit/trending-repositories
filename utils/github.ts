import { Octokit } from 'octokit';

export class Github {
  public itemsToFetch = 16;

  public async searchRepositories() {
    try {
      const octokit = new Octokit();
      const sevenDaysAgo: Date = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);
      const formattedDate: string = sevenDaysAgo.toISOString().split('T')[0];

      const response = await octokit.request('GET /search/repositories', {
        sort: 'stars',
        order: 'desc',
        q: `created:>${formattedDate}`,
        per_page: this.itemsToFetch,
        page: 1,
      });

      return response.data;
    } catch (error) {
      if (error instanceof Error) {
        console.error(error);
      }
    }
  }

  public async getTrendingRepositories(count?: number) {
    if (count) {
      this.itemsToFetch = count;
    }

    const trendingRepositories = await this.searchRepositories();
    const repositoriesDataMapped = trendingRepositories?.items.map((item) => {
      return {
        name: item.name,
        description: item.description,
        starCount: item.stargazers_count,
        id: item.id,
        language: item?.language,
        link: item.html_url,
      };
    });
    return repositoriesDataMapped;
  }
}
