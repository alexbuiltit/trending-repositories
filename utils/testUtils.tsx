import React, { ReactNode } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@/styles/theme';

interface ProviderProps {
  children: ReactNode;
}

const AllTheProviders = ({ children }: ProviderProps) => {
  return <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>;
};

const customRender = (
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  options?: RenderOptions<typeof import('@testing-library/dom/types/queries'), HTMLElement, HTMLElement>,
) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
