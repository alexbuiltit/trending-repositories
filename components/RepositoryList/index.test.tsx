import { RepositoryListType } from '@/types/types';
import { render, screen } from '@/utils/testUtils';
import { RepositoryList } from '.';

const repositoryListProps: RepositoryListType = {
  items: [
    {
      name: 'github-trending',
      description: 'A repo to display the trending items in GitHub',
      starCount: 1,
      language: 'TypeScript',
      id: 42,
      favourite: false,
      link: 'https://github.com/alexbuiltit',
    },
    {
      name: 'another-repo',
      description: 'An example repository',
      starCount: 1,
      language: 'Java',
      id: 4910,
      favourite: false,
      link: 'https://github.com',
    },
    {
      name: 'just-one-more',
      description: 'The final example repository',
      starCount: 1,
      language: 'Go',
      id: 4911120,
      favourite: false,
      link: 'https://github.com',
    },
  ],
  filters: [],
};

describe('RepositoryList', () => {
  it('should render a list of repositories as expected', () => {
    const { container } = render(<RepositoryList {...repositoryListProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should only show a list of TypeScript repositories', () => {
    repositoryListProps.filters = ['TypeScript'];
    const { container } = render(<RepositoryList {...repositoryListProps} />);
    expect(screen.queryByText('Java')).not.toBeInTheDocument();
    expect(screen.queryByText('Go')).not.toBeInTheDocument();
    expect(screen.queryAllByText('TypeScript').length).toEqual(1);
    expect(container).toMatchSnapshot();
  });
});
