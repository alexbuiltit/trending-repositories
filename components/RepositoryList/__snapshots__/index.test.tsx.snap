// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryList should only show a list of TypeScript repositories 1`] = `
<div>
  <ul
    class="sc-457fc63c-0 gcrsTl"
  >
    <li
      class="sc-c6d6bd61-0 fiYkvw"
      data-repo-id="42"
      data-testid="repository-card"
    >
      <div
        class="sc-c6d6bd61-1 hNhUMl"
      >
        <h2
          class="sc-c6d6bd61-3 cpSoRL"
        >
          <a
            href="https://github.com/alexbuiltit"
            rel="noreferrer"
            target="_blank"
          >
            github-trending
          </a>
        </h2>
        <span
          class="sc-1762ab62-0 eLoAgg"
        >
          TypeScript
        </span>
        <p>
          A repo to display the trending items in GitHub
        </p>
        <div
          class="sc-c6d6bd61-2 gychjE"
        >
          <p>
            Stars: 
            1
          </p>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`RepositoryList should render a list of repositories as expected 1`] = `
<div>
  <ul
    class="sc-457fc63c-0 gcrsTl"
  >
    <li
      class="sc-c6d6bd61-0 fiYkvw"
      data-repo-id="42"
      data-testid="repository-card"
    >
      <div
        class="sc-c6d6bd61-1 hNhUMl"
      >
        <h2
          class="sc-c6d6bd61-3 cpSoRL"
        >
          <a
            href="https://github.com/alexbuiltit"
            rel="noreferrer"
            target="_blank"
          >
            github-trending
          </a>
        </h2>
        <span
          class="sc-1762ab62-0 eLoAgg"
        >
          TypeScript
        </span>
        <p>
          A repo to display the trending items in GitHub
        </p>
        <div
          class="sc-c6d6bd61-2 gychjE"
        >
          <p>
            Stars: 
            1
          </p>
        </div>
      </div>
    </li>
    <li
      class="sc-c6d6bd61-0 fiYkvw"
      data-repo-id="4910"
      data-testid="repository-card"
    >
      <div
        class="sc-c6d6bd61-1 hNhUMl"
      >
        <h2
          class="sc-c6d6bd61-3 cpSoRL"
        >
          <a
            href="https://github.com"
            rel="noreferrer"
            target="_blank"
          >
            another-repo
          </a>
        </h2>
        <span
          class="sc-1762ab62-0 eLoAgg"
        >
          Java
        </span>
        <p>
          An example repository
        </p>
        <div
          class="sc-c6d6bd61-2 gychjE"
        >
          <p>
            Stars: 
            1
          </p>
        </div>
      </div>
    </li>
    <li
      class="sc-c6d6bd61-0 fiYkvw"
      data-repo-id="4911120"
      data-testid="repository-card"
    >
      <div
        class="sc-c6d6bd61-1 hNhUMl"
      >
        <h2
          class="sc-c6d6bd61-3 cpSoRL"
        >
          <a
            href="https://github.com"
            rel="noreferrer"
            target="_blank"
          >
            just-one-more
          </a>
        </h2>
        <span
          class="sc-1762ab62-0 eLoAgg"
        >
          Go
        </span>
        <p>
          The final example repository
        </p>
        <div
          class="sc-c6d6bd61-2 gychjE"
        >
          <p>
            Stars: 
            1
          </p>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
